encoder.go

```diff
@@ -0,0 +1,252 @@
+			nodeTypeFormatted := strings.Join(sliceFormat(nodeType, e.Formatters), ".")
+			w[nodeTypeFormatted] = f.Type().Name()
+		}
+		croots := roots
+		if len(roots) == 0 {
+			croots = append(roots, f.Type().Name())
+		}
+		if err := e.fdumpStruct(w, f, croots); err != nil {
+			return err
+		}
+	case reflect.Array, reflect.Slice:
+		if e.ExtraFields.Type {
+			nodeType := append(roots, "__Type__")
+			nodeTypeFormatted := strings.Join(sliceFormat(nodeType, e.Formatters), ".")
+			w[nodeTypeFormatted] = "Array"
+		}
+		if err := e.fDumpArray(w, i, roots); err != nil {
+			return err
+		}
+		return nil
+	case reflect.Map:
+		if e.ExtraFields.Type {
+			nodeType := append(roots, "__Type__")
+			nodeTypeFormatted := strings.Join(sliceFormat(nodeType, e.Formatters), ".")
+			w[nodeTypeFormatted] = "Map"
+		}
+		if err := e.fDumpMap(w, i, roots); err != nil {
+			return err
+		}
+		return nil
+	default:
+		k := fmt.Sprintf("%s", strings.Join(sliceFormat(roots, e.Formatters), "."))
+		w[k] = f.Interface()
+	}
+	return nil
+}
+
+func (e *Encoder) fDumpArray(w map[string]interface{}, i interface{}, roots []string) error {
+	v := reflect.ValueOf(i)
+
+	if e.ExtraFields.Len {
+		nodeLen := append(roots, "__Len__")
+		nodeLenFormatted := strings.Join(sliceFormat(nodeLen, e.Formatters), ".")
+		w[nodeLenFormatted] = v.Len()
+	}
+
+	for i := 0; i < v.Len(); i++ {
+		var l string
+		//croots := roots
+		var croots []string
+		if len(roots) > 0 {
+			l = roots[len(roots)-1:][0]
+			croots = append(roots, fmt.Sprintf("%s%d", l, i))
+		} else {
+			croots = append(roots, fmt.Sprintf("%s%d", l, i))
+		}
+		f := v.Index(i)
+		if err := e.fdumpInterface(w, f.Interface(), croots); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *Encoder) fDumpMap(w map[string]interface{}, i interface{}, roots []string) error {
+	v := reflect.ValueOf(i)
+
+	keys := v.MapKeys()
+
+	if e.ExtraFields.Len {
+		nodeLen := append(roots, "__Len__")
+		nodeLenFormatted := strings.Join(sliceFormat(nodeLen, e.Formatters), ".")
+		w[nodeLenFormatted] = len(keys)
+	}
+
+	for _, k := range keys {
+		key := fmt.Sprintf("%v", k.Interface())
+		croots := append(roots, key)
+		value := v.MapIndex(k)
+
+		f := valueFromInterface(value.Interface())
+
+		if validAndNotEmpty(f) && f.Type().Kind() == reflect.Struct {
+			croots = append(croots, f.Type().Name())
+		}
+
+		if err := e.fdumpInterface(w, value.Interface(), croots); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *Encoder) fdumpStruct(w map[string]interface{}, s reflect.Value, roots []string) error {
+	if e.ExtraFields.DetailedStruct {
+		if e.ExtraFields.Len {
+			nodeLen := append(roots, "__Len__")
+			nodeLenFormatted := strings.Join(sliceFormat(nodeLen, e.Formatters), ".")
+			w[nodeLenFormatted] = s.NumField()
+		}
+
+		structKey := fmt.Sprintf("%s", strings.Join(sliceFormat(roots, e.Formatters), "."))
+		if s.CanInterface() {
+			w[structKey] = fmt.Sprintf("%+v", s.Interface())
```

Comment:
@fsamin The string representation of the struct won't be very helpful, we need the concrete type.

Reply:
i hope c43ed88 will suit your needs

Reply:
LGTM. If you still want to have the `"%+v"` notation when dumping to a map of string, a special condition could be added based on the `reflect.Kind()` of the value.